Common principles: More elaborate "entity" definition

Lestropie opened the pull request:
Relates to bids-standard/bids-specification#947.

This is an attempt at the proposal I made in [this comment](https://github.com/bids-standard/bids-specification/pull/947#issuecomment-1005387248). There is however a little bit of a modification. As shown in [the poll](https://github.com/bids-standard/bids-specification/pull/947#issuecomment-998397213), there is no consensus around exactly what it is that is referred to as an "entity". In my comment, I had suggested that "entity" refers to a higher-order concept, of which things like the "name" (e.g. "reconstruction") and the "key" (e.g. "`rec`") are attributes. I had proposed that any and all references to such attributes would need to be explicit regarding exactly to what was being referred. However I found this to be quite clumsy. The proposal here therefore includes the following comment:

> Depending on context, any one of the entity name, key, format, or a specific
> entity instance, may in some instances be referred to as simply an "entity".

This means that most of the proposed changes in bids-standard/bids-specification#947 relating to the separation between a key-value entry in a JSON and an entity in a file name remain unchanged; this particular PR only addresses the definition of "entity".

-----

Note that, rather than simply including "entity" in the list of definitions at the commencement of "common principles", I have instead given it its own Level 2 heading. This is such a core concept to BIDS that I believe it warrants greater precedence than other terms that necessitate definition; moreover, these nuances about the various attributes of entities and the way they are described requires a greater volume of text than is required for the other defined terms. There are some surrounding changes necessary because the dictionary terms are defined before the "Entity" heading; happy to separate these out if maintainers see fit, this is just the initial set of changes I deemed necessary to construct the definition as desired.

Lestropie committed:
Common principles: More elaborate "entity" definition

sappelhoff commented on src/02-common-principles.md:
```diff
@@ -105,36 +102,62 @@ misunderstanding we clarify them here.
+it appears.
+
+Depending on context, any one of the entity name, key, format, or a specific
+entity instance, may in some instances be referred to as simply an "entity".
```
```suggestion
entity instance, may be referred to as simply an "entity".
```

Lestropie replied:
Closed by 3972dd7.

sappelhoff commented on src/02-common-principles.md:
```diff
@@ -105,36 +102,62 @@ misunderstanding we clarify them here.
+above are all examples of entities. The comprehensive list of supported
+entities is defined in [Appendix IX](99-appendices/09-entities.md);
+further, whether each is OPTIONAL, REQUIRED, or MUST NOT be provided for
+various data files, as well as their relative ordering in a file name, are
+defined in the Entity Table in [Appendix IV](99-appendices/04-entity-table.md).
```
line breaks and added a word "compulsory"

```suggestion
various data files, as well as their compulsory relative ordering in a file name,
are defined in the Entity Table in
[Appendix IV](99-appendices/04-entity-table.md).
```

Lestropie replied:
Closed by 3972dd7.

sappelhoff commented on src/02-common-principles.md:
```diff
@@ -153,50 +176,55 @@ The solutions will change from case to case and publicly available datasets will
+If data for the subject were acquired across multiple sessions, then within
+the subject directory resides sub-directories named "`ses-<label>`",
+where string "`<label>`" is substituted with a unique identification
+label for each session. In datasets where at least one subject has more than
+one session, this additional sub-directory later SHOULD be added for all
+subjects in the dataset. Additional information on each session MAY be
```
I know this recommendation was in the spec prior to this PR, but I don't really get the point: What if subjects miss some sessions ... or if it's a between subject design with different subjects being part of different sessions?

See also the recommendation below:

> A data type directory SHOULD NOT be defined if
there are no files to be placed in that directory.

We don't have to discuss it here, but if someone has a quick point on this, I am happy to hear it.

Lestropie replied:
My suspicion is that it was intended to make things easier for App developers. If a hypothetical App is hard-wired to loop over multiple sessions within a subject and process each individually, but without adequate logic to detect the absence of a session structure in any particular subject, passing such a dataset could break the App. Personally I'm happy to have them break; but I'm a known software sadist so that might not count for much.

sappelhoff reviewed (approved):
Thanks for this very clear PR @Lestropie -- I think your added definitions of different aspects of an entity improve the situation, and being lenient with what to call "an entity" (the instance, the key, ...) will probably work well with the situation you identified with your poll in https://github.com/bids-standard/bids-specification/pull/947

I left a few minor comments. 
One additional comment: Could you please insert line breaks after the end of each sentence, and meaningfully after "semantic break points" in a sentence?
That will probably help us to get a cleaner git diff if some particular sentence needs to be changed down the line.

Lestropie committed:
Common principles: Reformmating of Entity definition changes

Requests made in: Lestropie/bids-specification#4.
Note that this is a repeat of commit 3972dd7d, but with corrected multi-author attribution.

Co-authored-by: Stefan Appelhoff <stefan.appelhoff@mailbox.org>

Lestropie committed:
Common principles: Remark fixes
